feat: Add jsonnet-bundler support

andrein opened the pull request:
## Changes:

This PR adds support for [jsonnet-bundler](https://github.com/jsonnet-bundler/jsonnet-bundler).

TODO:
- [X] implement `extractPackageFile()`
- [x] implement `updateArtifacts()`
- [x] tests
- [x] docs
- [x] https://github.com/containerbase/buildpack/pull/196

## Context:

Closes #9800 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

rarkins commented on lib/manager/jsonnet-bundler/extract.spec.ts:
```diff
@@ -0,0 +1,47 @@
+const jsonnetfile = loadFixture('jsonnetfile.json');
+const jsonnetfileWithName = loadFixture('jsonnetfile-with-name.json');
+
+describe('manager/jsonnet-bundler', () => {
```
```suggestion
describe('manager/jsonnet-bundler/extract', () => {
```

viceice commented on lib/manager/jsonnet-bundler/artifacts.spec.ts:
```diff
@@ -0,0 +1,38 @@
+
+describe('manager/jsonnet-bundler/artifacts', () => {
+  beforeEach(() => {
+    jest.resetAllMocks();
+    jest.resetModules();
```
```suggestion
```

I don't think this is needed.

viceice commented on lib/manager/jsonnet-bundler/artifacts.spec.ts:
```diff
@@ -0,0 +1,38 @@
+
+    setGlobalConfig(adminConfig);
+  });
+  afterEach(() => {
+    setGlobalConfig();
+  });
```
```suggestion
```

not needed.

viceice commented on lib/manager/jsonnet-bundler/artifacts.ts:
```diff
@@ -0,0 +1,91 @@
+import { readLocalFile } from '../../util/fs';
+import { getRepoStatus } from '../../util/git';
+import { regEx } from '../../util/regex';
+import {
```
```suggestion
import type {
```

viceice commented on lib/manager/jsonnet-bundler/artifacts.ts:
```diff
@@ -0,0 +1,91 @@
+
+function dependencyUrl(dep: PackageDependency): string {
+  const url = dep.lookupName;
+  if (dep.managerData.subdir) {
```
why not add `subdir` to `lookupName`? Have you tried it?

andrein replied:
Tried it just now, it doesn't resolve the dependency properly:
```
DEBUG: packageFiles with updates (repository=andrein/renovate-test)
       "config": {
         "jsonnet-bundler": [
           {
             "packageFile": "jsonnet-bundler/jsonnetfile.json",
             "deps": [
               {
                 "depName": "thanos",
                 "lookupName": "https://github.com/thanos-io/thanos.git/mixin",
                 "currentValue": "v0.22.0",
                 "datasource": "git-tags",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "thanos",
                     "message": "Failed to look up dependency thanos"
                   }
                 ],
                 "versioning": "semver"
               }
             ]
           }
         ],
```

viceice commented on lib/manager/jsonnet-bundler/extract.spec.ts:
```diff
@@ -0,0 +1,68 @@
+import { loadFixture } from '../../../test/util';
+import { extractPackageFile } from './extract';
```
```suggestion
import { extractPackageFile } from '.';
```

coverage

viceice commented on lib/manager/jsonnet-bundler/artifacts.spec.ts:
```diff
@@ -0,0 +1,38 @@
+import { RepoGlobalConfig } from '../../config/types';
+import * as docker from '../../util/exec/docker';
+import { UpdateArtifactsConfig } from '../types';
+import { updateArtifacts } from './artifacts';
```
```suggestion
import { updateArtifacts } from '.';
```

coverage

viceice commented on lib/manager/jsonnet-bundler/artifacts.ts:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  try {
+    let cmd;
```
```suggestion
    let cmd: string;
```

viceice commented on lib/manager/jsonnet-bundler/extract.ts:
```diff
@@ -0,0 +1,62 @@
+import { GitTagsDatasource } from '../../datasource/git-tags';
+import { logger } from '../../logger';
+import { regEx } from '../../util/regex';
+import { PackageDependency, PackageFile } from '../types';
+import { Dependency, JsonnetFile } from './types';
```
```suggestion
import type { PackageDependency, PackageFile } from '../types';
import type { Dependency, JsonnetFile } from './types';
```

viceice commented on lib/manager/jsonnet-bundler/extract.ts:
```diff
@@ -0,0 +1,62 @@
+  content: string,
+  packageFile: string
+): PackageFile | null {
+  logger.trace(`jsonnet-bundler.extractPackageFile(${packageFile})`);
+  logger.trace({ content });
```
```suggestion
  logger.trace({packageFile}, 'jsonnet-bundler.extractPackageFile()');
```

viceice commented on lib/manager/jsonnet-bundler/extract.ts:
```diff
@@ -0,0 +1,62 @@
+  const deps: PackageDependency[] = [];
+  let jsonnetFile: JsonnetFile;
+  try {
+    jsonnetFile = JSON.parse(content);
```
```suggestion
    jsonnetFile = JSON.parse(content) as JsonnetFile;
```

viceice commented on lib/manager/jsonnet-bundler/extract.ts:
```diff
@@ -0,0 +1,62 @@
+  return { deps };
+}
+
+export function extractDependency(
```
Don't export if  there is no public use.

andrein replied:
```suggestion
function extractDependency(
```

viceice commented on lib/manager/jsonnet-bundler/index.ts:
```diff
@@ -0,0 +1,9 @@
+import { updateArtifacts } from './artifacts';
+import { extractPackageFile } from './extract';
```
```suggestion
export { updateArtifacts } from './artifacts';
export { extractPackageFile } from './extract';
```

andrein commented:
The only remaining thing on my todo is "add jb to https://github.com/containerbase/buildpack". How do I do that?

viceice commented:
Marking as draft, as it needs a buildpack update first.

viceice commented:
> The only remaining thing on my todo is "add jb to https://github.com/containerbase/buildpack". How do I do that?

Please open an issue there to discuss. Add some references how to install `jb` (should also be possible without root)

viceice commented on lib/manager/jsonnet-bundler/artifacts.spec.ts:
```diff
@@ -0,0 +1,200 @@
+import { exec as _exec } from 'child_process';
```
```suggestion
```

import from `exec-util`. 🙃

viceice commented on lib/manager/jsonnet-bundler/artifacts.spec.ts:
```diff
@@ -0,0 +1,200 @@
+const exec: jest.Mock<typeof _exec> = _exec as any;
+const env = mocked(_env);
+jest.mock('child_process');
+jest.mock('../../util/exec/env');
+jest.mock('../../../lib/util/fs');
+jest.mock('../../../lib/util/git');
```
Some import are wrong here
```suggestion
jest.mock('../../util/exec/env');
jest.mock('../../util/fs');
jest.mock('../../util/git');
```

viceice commented on lib/manager/jsonnet-bundler/artifacts.spec.ts:
```diff
@@ -0,0 +1,200 @@
+import { exec as _exec } from 'child_process';
+import { join } from 'upath';
+import { envMock, mockExecAll } from '../../../test/exec-util';
+import { fs, git, mocked } from '../../../test/util';
+import { setGlobalConfig } from '../../config/global';
+import { RepoGlobalConfig } from '../../config/types';
+import * as _env from '../../util/exec/env';
+import { StatusResult } from '../../util/git';
+import { UpdateArtifactsConfig } from '../types';
```
```suggestion
import { envMock, exec, mockExecAll } from '../../../test/exec-util';
import { env, fs, git, mocked } from '../../../test/util';
import { setGlobalConfig } from '../../config/global';
import type { RepoGlobalConfig } from '../../config/types';
import type { StatusResult } from '../../util/git';
import type { UpdateArtifactsConfig } from '../types';
```

viceice commented on lib/manager/jsonnet-bundler/artifacts.spec.ts:
```diff
@@ -0,0 +1,200 @@
+import { UpdateArtifactsConfig } from '../types';
+import { updateArtifacts } from '.';
+
+const exec: jest.Mock<typeof _exec> = _exec as any;
+const env = mocked(_env);
```
```suggestion
```

viceice commented on lib/manager/jsonnet-bundler/extract.ts:
```diff
@@ -0,0 +1,59 @@
+      dependency.name || match.groups.depName || dependency.source.git.remote,
+    lookupName: dependency.source.git.remote,
+    currentValue: dependency.version,
+    datasource: GitTagsDatasource.id,
```
move this to `defaultConfig` if it's same for all deps

viceice commented on lib/manager/jsonnet-bundler/artifacts.ts:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  try {
+    const execOptions: ExecOptions = {
```
I think we need to extract the `version` so we can use the right `jb` version.

https://github.com/jsonnet-bundler/jsonnet-bundler/releases/tag/v0.3.1

andrein replied:
I don't think so. The release notes state that it's v0.3.1 can handle the jsonnetfile.json files produced by 0.2.0.
```
The format of jsonnetfile.json has changed. While v0.3.0 can
handle the old v0.2 format, v0.2 can't and must not be used with a
jsonnetfile.json created by v0.3.0
```

I'm targeting jsonnet-bundler 0.4.0 anyway, since starting from that version you can update single dependencies. In pre-4.0.0, `jb update` would update all dependencies uncoditionally.

viceice replied:
Yes, please note this in the readme.

viceice commented on lib/manager/jsonnet-bundler/artifacts.ts:
```diff
@@ -0,0 +1,102 @@
+
+  try {
+    const execOptions: ExecOptions = {
+      cwdFile: packageFileName,
```
```suggestion
      cwdFile: packageFileName,
      docker: {
         image: 'sidecar',
      },
```
Please add a `ToolConstraint` for jb like composer. If we can't extract the version somehow, simply pass `config.constraints?.jb`.

https://github.com/renovatebot/renovate/blob/4e07ddf52652445024a78a04dd70bc34d6be4d19/lib/manager/composer/artifacts.ts#L99-L121

https://github.com/renovatebot/renovate/blob/04620d71a8650e742e2eb3bbbf0e778b63554f1d/lib/util/exec/buildpack.ts#L8-L14

viceice commented on lib/manager/jsonnet-bundler/artifacts.ts:
```diff
@@ -0,0 +1,110 @@
+import { TEMPORARY_ERROR } from '../../constants/error-messages';
+import { logger } from '../../logger';
+import { ExecOptions, exec } from '../../util/exec';
+import { ToolConstraint } from '../../util/exec/types';
```
```suggestion
import type { ToolConstraint } from '../../util/exec/types';
```

viceice reviewed (commented):
can you please do a real run with `binarySource=docker` ?

andrein commented:
I did a real run with binarySource=docker and it failed because the sidecar image isn't published yet. I've inspected the output and it looks like it's doing the right thing (tm)

```DEBUG: rawExec err (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"",
         "stdout": "",
         "stderr": "No tool defined - skipping: jb\n",
         "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"\nNo tool defined - skipping: jb\n",
         "stack": "Error: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"\nNo tool defined - skipping: jb\n\n    at ChildProcess.exithandler (node:child_process:397:12)\n    at ChildProcess.emit (node:events:390:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1062:16)\n    at Socket.<anonymous> (node:internal/child_process:448:11)\n    at Socket.emit (node:events:390:28)\n    at Socket.emit (node:domain:475:12)\n    at Pipe.<anonymous> (node:net:687:12)"
       }```

viceice commented:
> I did a real run with binarySource=docker and it failed because the sidecar image isn't published yet. I've inspected the output and it looks like it's doing the right thing (tm)
> 
> ```
>        "err": {
>          "killed": false,
>          "code": 1,
>          "signal": null,
>          "cmd": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"",
>          "stdout": "",
>          "stderr": "No tool defined - skipping: jb\n",
>          "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"\nNo tool defined - skipping: jb\n",
>          "stack": "Error: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"\nNo tool defined - skipping: jb\n\n    at ChildProcess.exithandler (node:child_process:397:12)\n    at ChildProcess.emit (node:events:390:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1062:16)\n    at Socket.<anonymous> (node:internal/child_process:448:11)\n    at Socket.emit (node:events:390:28)\n    at Socket.emit (node:domain:475:12)\n    at Pipe.<anonymous> (node:net:687:12)"
>        }```
> ```

renovate is missing on https://github.com/renovatebot/docker-sidecar, so the image isn't yet updated with latest buildpack [v2.3.0](https://github.com/containerbase/buildpack/releases/tag/2.3.0)

andrein commented:
Shall I make a PR to bump the version/add renovate?

viceice commented:
> Shall I make a PR to bump the version/add renovate?

nope, the app simply wasn't installed on that repo, which is now fixed and the update should be release in a couple of minutes

viceice commented:
ok, released: https://hub.docker.com/r/renovate/sidecar/tags

Please retry docker binary source

andrein commented:
Retried, and it failed with: 

```
DEBUG: Executing command (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
       "command": [
         "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\""
       ]
DEBUG: rawExec err (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
       "err": {
         "killed": false,
         "code": 1,
         "signal": null,
         "cmd": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"",
         "stdout": "Installing tool jb v0.4.0\n",
         "stderr": "v0.4.0\njb: error: failed to load jsonnetfile: open /var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler/jsonnetfile.json: no such file or directory\n",
         "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"\nv0.4.0\njb: error: failed to load jsonnetfile: open /var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler/jsonnetfile.json: no such file or directory\n",
         "stack": "Error: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"\nv0.4.0\njb: error: failed to load jsonnetfile: open /var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler/jsonnetfile.json: no such file or directory\n\n    at ChildProcess.exithandler (node:child_process:397:12)\n    at ChildProcess.emit (node:events:390:28)\n    at ChildProcess.emit (node:domain:475:12)\n    at maybeClose (node:internal/child_process:1062:16)\n    at Socket.<anonymous> (node:internal/child_process:448:11)\n    at Socket.emit (node:events:390:28)\n    at Socket.emit (node:domain:475:12)\n    at Pipe.<anonymous> (node:net:687:12)"
       }
```

Did I miss anything?

viceice commented:
Your local user id is probably not 1000, so the docker sidecar can't read the git checkout.

HonkingGoose commented on lib/manager/jsonnet-bundler/readme.md:
```diff
@@ -0,0 +1,5 @@
+
+Supports [lock file maintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance).
+
+This plugin requires `jsonnet-bundler >= v0.4.0` since previous versions don't support updating single dependencies.
```
The documentation says you should use `jsonnet-bundler` equal or greater than `v0.4.0`.

I searched for the string `0.4.0` in the changed files, and get no hits.

Should we have code on the Renovate side that checks the version of `jsonnet-bundler`, or can we rely on the bundler itself to throw an error when you use a unsupported version?

andrein replied:
@HonkingGoose see [this](https://github.com/renovatebot/renovate/pull/12720/files/05546caad318812c35bc76b594cb5273f7233ce2#diff-bfabdbd781a38490cf7b487876d0c3de22611b32dbf1ff471ef51cbc5c976d3eR40) review for context.

jb 0.4.0 was added in buildpack, but now that I think of it, if you're not runnig renovate with the docker binarySource you will get an error like this (jb 0.2.0 in this case):
```
❯ ./jb-darwin-amd64 update https://github.com/thanos-io/thanos/jsonnet/mixin
Error parsing commandline arguments: unexpected https://github.com/thanos-io/thanos/jsonnet/mixin
jb-darwin-amd64: error: unexpected https://github.com/thanos-io/thanos/jsonnet/mixin
```

I can add a version check if this is something you think this should be handled more gracefully. Is there another manager I could draw inspiration from?

rarkins replied:
We don't usually check for the minimum supported versions of binaries when binarySource=global, so I don't think we need to do this here either

andrein commented:
> 

That wasn't it. I was connecting to a remote docker daemon and volumes weren't working (*suprise*).

When I ran it with a local docker daemon it worked:

```
DEBUG: No tag or tagConstraint specified (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
       "image": "docker.io/renovate/sidecar"
DEBUG: Fetching Docker image: docker.io/renovate/sidecar (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
DEBUG: Finished fetching Docker image (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
DEBUG: Executing command (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
       "command": [
         "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\""
       ]
DEBUG: exec completed (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
       "cmd": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test\" -v \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\":\"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/cache\" -w \"/var/folders/_t/f_q95xcd0ws9s_cqvh6tpgfh0000gn/T/renovate/repos/github/andrein/renovate-test/jsonnet-bundler\" docker.io/renovate/sidecar bash -l -c \"install-tool jb v0.4.0 && jb update https://github.com/thanos-io/thanos.git/mixin\"",
       "durationMs": 9776,
       "stdout": "Installing tool jb v0.4.0\n",
       "stderr": "v0.4.0\nGET https://github.com/thanos-io/thanos/archive/68ff28df232fc1f610a04b114290bdb68d1ba61e.tar.gz 200\n"
DEBUG: Updated 1 package files (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
DEBUG: Updated 11 lock files (repository=andrein/renovate-test, branch=renovate/thanos-0.x)
       "updatedArtifacts": [
         "jsonnet-bundler/jsonnetfile.json",
         "jsonnet-bundler/jsonnetfile.lock.json",
         "jsonnet-bundler/vendor/github.com/thanos-io/thanos/mixin/README.md",
         "jsonnet-bundler/vendor/github.com/thanos-io/thanos/mixin/dashboards/bucket_replicate.libsonnet",
         "jsonnet-bundler/vendor/github.com/thanos-io/thanos/mixin/dashboards/compact.libsonnet",
         "jsonnet-bundler/vendor/github.com/thanos-io/thanos/mixin/dashboards/query.libsonnet",
         "jsonnet-bundler/vendor/github.com/thanos-io/thanos/mixin/dashboards/receive.libsonnet",
         "jsonnet-bundler/vendor/github.com/thanos-io/thanos/mixin/dashboards/rule.libsonnet",
         "jsonnet-bundler/vendor/github.com/thanos-io/thanos/mixin/dashboards/sidecar.libsonnet",
         "jsonnet-bundler/vendor/github.com/thanos-io/thanos/mixin/dashboards/store.libsonnet",
         "jsonnet-bundler/vendor/github.com/thanos-io/thanos/mixin/lib/thanos-grafana-builder/README.md"
       ]
```

HonkingGoose commented on docs/usage/configuration-options.md:
```diff
@@ -1198,7 +1198,7 @@ With the above config, every PR raised by Renovate will have the label `dependen
 This feature can be used to refresh lock files and keep them up-to-date.
 "Maintaining" a lock file means recreating it so that every dependency version within it is updated to the latest.
-Supported lock files are `package-lock.json`, `yarn.lock`, `composer.lock`, `Gemfile.lock`, `poetry.lock` and `Cargo.lock`.
+Supported lock files are `package-lock.json`, `yarn.lock`, `composer.lock`, `Gemfile.lock`, `poetry.lock`, `Cargo.lock` and `jsonnetfile.lock.json`.
```
```suggestion
Supported lock files are:

- `package-lock.json`
- `yarn.lock`
- `composer.lock`
- `Gemfile.lock`
- `poetry.lock`
- `Cargo.lock`
- `jsonnetfile.lock.json`

```

I think this will look better as a list, and be easier to read as well.

HonkingGoose reviewed (approved):
Documentation looks good to me!
